File: src/material.rs
use crate::ray::Ray;
use crate::rtweekend::{random_double, random_in_unit_sphere, random_unit_vector};
use crate::{color, vec3, Color, HitRecord};

pub trait Material {
    fn scatter(
        &self,
        r_in: &Ray,
        rec: &HitRecord,
        attenuation: &mut Color,
        scattered: &mut Ray,
    ) -> bool;
}

pub struct Lambertian {
    pub albedo: Color,
}

impl Lambertian {
    pub fn new(albedo: &Color) -> Self
    where
        Self: Sized,
    {
        Lambertian {
            albedo: albedo.clone(),
        }
    }
}

impl Material for Lambertian {
    fn scatter(
        &self,
        _r_in: &Ray,
        rec: &HitRecord,
        attenuation: &mut Color,
        scattered: &mut Ray,
    ) -> bool {
        let mut scatter_direction = rec.normal + random_unit_vector();

        // Catch degenerate scatter direction
        if scatter_direction.near_zero() {
            scatter_direction = rec.normal;
        }

        *scattered = Ray::new(&rec.p, &scatter_direction);
        *attenuation = self.albedo;

        true
    }
}

pub struct Metal {
    pub albedo: Color,
    pub fuzz: f64,
}

impl Metal {
    pub fn new(albedo: &Color, f: f64) -> Self
    where
        Self: Sized,
    {
        Metal {
            albedo: albedo.clone(),
            fuzz: if f < 1.0 { f } else { 1.0 },
        }
    }
}

impl Material for Metal {
    fn scatter(
        &self,
        r_in: &Ray,
        rec: &HitRecord,
        attenuation: &mut Color,
        scattered: &mut Ray,
    ) -> bool {
        let reflected = r_in.direction.unit_vector().reflect(&rec.normal);
        *scattered = Ray::new(&rec.p, &(reflected + self.fuzz * random_in_unit_sphere()));
        *attenuation = self.albedo;

        scattered.direction.dot(&rec.normal) > 0.0
    }
}

pub struct Dielectric {
    /// Index of Refraction
    pub ir: f64,
}

impl Dielectric {
    pub fn new(index_of_refraction: f64) -> Self
    where
        Self: Sized,
    {
        Dielectric {
            ir: index_of_refraction,
        }
    }

    /// Use Schlick's approximation for reflectance.
    fn reflectance(cosine: f64, ref_idx: f64) -> f64 {
        let mut r0 = (1.0 - ref_idx) / (1.0 + ref_idx);
        r0 *= r0;
        r0 + (1.0 - r0) * (1.0 - cosine).powi(5)
    }
}

impl Material for Dielectric {
    fn scatter(
        &self,
        r_in: &Ray,
        rec: &HitRecord,
        attenuation: &mut Color,
        scattered: &mut Ray,
    ) -> bool {
        *attenuation = color![1.0, 1.0, 1.0];
        let refraction_ratio = if rec.front_face {
            1.0 / self.ir
        } else {
            self.ir
        };

        let unit_direction = r_in.direction.unit_vector();
        let cos_theta = (-unit_direction).dot(&rec.normal).min(1.0);
        let sin_theta = (1.0 - cos_theta * cos_theta).sqrt();

        let cannot_refract = refraction_ratio * sin_theta > 1.0;
        let direction = if cannot_refract
            || Dielectric::reflectance(cos_theta, refraction_ratio) > random_double()
        {
            unit_direction.reflect(&rec.normal)
        } else {
            unit_direction.refract(&rec.normal, refraction_ratio)
        };

        *scattered = Ray::new(&rec.p, &direction);

        true
    }
}

File: tj/material.py
from enum import IntEnum

import taichi as ti
from taichi.math import *

from ray import Ray
from rtweekend import random_unit_vector, near_zero, random_in_unit_sphere, refract_


class MATERIAL(IntEnum):
    NONE = 0
    LAMBERTIAN = 1
    METAL = 2
    DIELECTRIC = 3


@ti.dataclass
class Material:
    typ: int = MATERIAL.NONE
    albedo: vec3 = vec3(0)
    fuzz: float = 0  # METAL: fuzz DIELECTRIC: ir

    @ti.func
    def scatter(self, ray_in: Ray, rec) -> (bool, vec3, Ray):
        r, a, s = False, vec3(0), Ray(vec3(0), vec3(0))
        if self.typ == MATERIAL.LAMBERTIAN:
            r, a, s = lambertian_scatter(ray_in, rec, self.albedo)
        elif self.typ == MATERIAL.METAL:
            r, a, s = metal_scatter(ray_in, rec, self.albedo, self.fuzz)
        elif self.typ == MATERIAL.DIELECTRIC:
            r, a, s = dielectric_scatter(ray_in, rec, self.fuzz)

        return r, a, s


def create_lambertian(albedo: vec3) -> Material:
    return Material(MATERIAL.LAMBERTIAN, albedo)


@ti.func
def lambertian_scatter(ray_in: Ray, rec, albedo) -> (bool, vec3, Ray):
    scatter_direction = rec.normal + random_unit_vector()
    if near_zero(scatter_direction):
        scatter_direction = rec.normal
    scattered = Ray(rec.p, scatter_direction)
    attenuation = albedo
    return True, attenuation, scattered


def create_metal(albedo: vec3, fuzz: float) -> Material:
    fuzz = min(fuzz, 1)
    return Material(MATERIAL.METAL, albedo, fuzz)


@ti.func
def metal_scatter(ray_in: Ray, rec, albedo, fuzz) -> (bool, vec3, Ray):
    reflected = reflect(normalize(ray_in.direction), rec.normal)
    scattered = Ray(rec.p, reflected + fuzz * random_in_unit_sphere())
    attenuation = albedo
    return dot(scattered.direction, rec.normal) > 0, attenuation, scattered


def create_dielectric(ir: float) -> Material:
    return Material(MATERIAL.DIELECTRIC, fuzz=ir)


@ti.func
def dielectric_scatter(ray_in: Ray, rec, ir) -> (bool, vec3, Ray):
    attenuation = vec3(1.0)
    refraction_ratio = ir if not rec.front_face else 1 / ir

    unit_direction = normalize(ray_in.direction)
    cos_theta = min(dot(-unit_direction, rec.normal), 1.0)
    sin_theta = ti.sqrt(1.0 - cos_theta * cos_theta)

    cannot_refract = refraction_ratio * sin_theta > 1.0
    direction = vec3(0)
    if cannot_refract or reflectance(cos_theta, refraction_ratio) > ti.random():
        direction = reflect(unit_direction, rec.normal)
    else:
        direction = refract_(unit_direction, rec.normal, refraction_ratio)

    scattered = Ray(rec.p, direction)
    return True, attenuation, scattered


@ti.func
def reflectance(cosine: float, ref_idx: float) -> float:
    r0 = (1 - ref_idx) / (1 + ref_idx)
    r0 = r0 * r0
    return r0 + (1 - r0) * ti.pow((1 - cosine), 5)
